Add texture grouping enginx eng diff UI

RichardWaiteSTFC opened the pull request:
**Description of work.**

Refactor of engineering diffraction interface - specifically calibration and focusing tabs.
The purpose was to support texture analysis (custom grouping of detectors into 20 sub-groups) - this required a significant re-write of the interface to generalise to nspectra (as opposed to max. 2 for both banks).

To this end the `CalibrationInfo` class has been expanded to handle the grouping to encapsulate the logic specific to different types of grouping and the generation of grouping files - this includes the use of an Enum to distinguish the allowed groupings. This removes a lot of duplicated code with length if/else statements and means that all types of groupings are treated in the same way. A lot of the `CalibrationInfo` methods make functions in the EngUtils library redundant - presently this library is kept as is so as not to interfere with the systems test `EnginXScriptTest` - this is because there is a separate issue to re-write the `EnginX.py` script using methods from the UI code #32663.

There is some change to the behaviour of the UI
1.  Calibration and focusing files are saved to a single directory (e.g. either /Focus or /User/RBnum/Focus) depending on whether an RB number is specified - this is because focusing the texture grouping writes ~40 files per run (20 in TOF and 20 in dSpacing) and duplicating this was affecting the run time.
2. Focusing now saves one .gss and .abc file per run in a given xunit (as opposed to one per spectrum) - this was requested by the instrument scientist.
3. Plotting focused data now produces one figure per run - with all spectra on the same axes - this was done to keep the number of plots reasonable for texture grouping.

**To test:**

(1) Follow the manual testing instructions (https://developer.mantidproject.org/Testing/EngineeringDiffraction/EngineeringDiffractionTestGuide.html)
(2) Try focusing a run with zero proton charge (326016)
(3) Focus and calibrate a run using the texture grouping
(4) Load the focused data into the fitting tab

Fixes #31905 and #31994 and #32290

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

RichardWaiteSTFC committed:
Add grouping file for texture analysis

RichardWaiteSTFC committed:
Add combined grouping for both banks in single xml

RichardWaiteSTFC committed:
Make new (unused) GroupInfo class

RichardWaiteSTFC committed:
Update CalibrationInfo class to inherit from GroupInfo and refactor

WORK IN PROGRESS!

RichardWaiteSTFC committed:
Fix creation of calibraito noutput files to work with new classes

RichardWaiteSTFC committed:
Run PDCal once for multiple groupings e.g. both, texture

Generalise extraction of diff consts for multiple groupings, remove unecessary clone of raw data (it is the focussed data that gets rebinned).
Added warning if PDCal failed for any spectra.

RichardWaiteSTFC committed:
Refactor plotting of output to support multiple groupings

RichardWaiteSTFC committed:
Fix path generation and file loading for existing calibrations

RichardWaiteSTFC committed:
Fix bugs on making new calibration

RichardWaiteSTFC committed:
Add cal table as attribute of CalibrationInfo

RichardWaiteSTFC committed:
Remove pending calibration object and avoid need to deepcopy

RichardWaiteSTFC committed:
Fix bug for plotting results of cropped calibration

RichardWaiteSTFC committed:
Write out separate .prm and cal table for both banks if not cropped

RichardWaiteSTFC committed:
Correctly generate prm and output for custom/cropped

RichardWaiteSTFC committed:
Merge branch 'master' into 31905_add_texture_grouping_ENGINX_EngDiffU

RichardWaiteSTFC committed:
Fix bugs for loading custom cropped calibration now no vanadium

RichardWaiteSTFC committed:
Load only one nexus per calibration for multiple banks

RichardWaiteSTFC committed:
Load full instrument calib on loading saved calibration

RichardWaiteSTFC committed:
Refactor focussing tab - operate on multi spectra workspaces

RichardWaiteSTFC committed:
Populate calibraiton fields when loading exisiting

Also create grouping workspace with instrument not MatrixWorkspace

RichardWaiteSTFC committed:
Make Fcus dirs if do not exist

RichardWaiteSTFC committed:
Check ADS for focused vanadium to avoid re-processing

RichardWaiteSTFC committed:
Fix bug so plot each spectra in different color

RichardWaiteSTFC committed:
Check ADS for raw vanadium in case just ROI changed

RichardWaiteSTFC committed:
Update curve label when plotting focussed output

RichardWaiteSTFC committed:
Remove unused method to write logs to file

RichardWaiteSTFC committed:
Copy focused van when rebinning in normalisation to preserve data

RichardWaiteSTFC committed:
Save output in d and TOF

Also removed unused variables from model

RichardWaiteSTFC committed:
Update last focussed files to pass onto fitting tab

RichardWaiteSTFC committed:
Change column type to str in fitting data to handle bankid change

RichardWaiteSTFC committed:
Bug fix - use run number of sample rather than calib in foc filename

RichardWaiteSTFC committed:
Merge branch 'master' into 31905_add_texture_grouping_ENGINX_EngDiffU

RichardWaiteSTFC committed:
Add texture grouping to cropping options

RichardWaiteSTFC committed:
Clear first calibration and convert to TOF before second calibration

Ensure focus tab follows same steps as calibration tab - and that the second calib is applied to vana

RichardWaiteSTFC committed:
Plot focused spectra in d-spacing

RichardWaiteSTFC committed:
Fix runtime error thrown when focusing is calib workspaces deleted

RichardWaiteSTFC committed:
Incorporate GroupInfo class into CalibrationInfo and remove GroupInfo

CalibrationInfo was just a wrapper for a lot of GroupInfo methods so can jsut have single class CalibrationInfo

RichardWaiteSTFC committed:
Merge branch 'master' into 31905_add_texture_grouping_ENGINX_EngDiffU

RichardWaiteSTFC committed:
Plot poly fit over multiple windows in texture

Also pass user supplied d-spacing so get maxiumum precision

RichardWaiteSTFC committed:
Refactor code to write diff consts to table to avoid extra loops

RichardWaiteSTFC committed:
Fix tests for cropping widget and fitting tab

RichardWaiteSTFC committed:
Write .prm using correct instrument params and Bk2Bk coef

Get Bk2Bk coef ffrom instrument tree rather than hard-coding North Bank/South bank so more general for different instruments.

RichardWaiteSTFC committed:
Write tests for calibration model

Calls to PDCal untested as this will be done inn systems test

RichardWaiteSTFC committed:
Remove prm template from calibrationInfo object now using only header

RichardWaiteSTFC committed:
Update last calibration with prm in presenter only on success

RichardWaiteSTFC committed:
Refactor calibration presenter tests

RichardWaiteSTFC committed:
Fix tests for focus presenter

RichardWaiteSTFC committed:
import func to load full calib from calib model in focus pres

RichardWaiteSTFC committed:
Put vanadium process in separate func in focus presenter

RichardWaiteSTFC committed:
Single ASCII files contianing all foc spectra and duplicate in RB dir

Save ASCII and .nxs files to a single directory, RB if specifed otherwise top level set in settings.

RichardWaiteSTFC committed:
Handle run or vanadium if have zero proton charge

Throw runtime error for latter

RichardWaiteSTFC committed:
Call SaveNexus with index list rather than use ExtractSpectrum

Fix bug where set bankid log twice per run and corrected fname of nxs files

RichardWaiteSTFC committed:
Rewrite focus model tests

RichardWaiteSTFC committed:
Update test file for IO/proj save with string bank id

RichardWaiteSTFC committed:
Assert save nxs using model last foc files

RichardWaiteSTFC committed:
Fix bug in test where mock obj interpreted as bool

RichardWaiteSTFC committed:
Mock out set_settigns that was writing to a conf file for linux

RichardWaiteSTFC committed:
Replace pathsep with semi-colon - same in linux and windows config

DannyHindson reviewed (changes requested):
I've had an initial look through and all looks pretty good, well done for getting this big change done. I've added a few detail level comments + here are a couple of more general questions\comments:

- at some point I think the focus plots on the engineering UI testing guide will need updating now you're plotting all the groups on a single set of axes
- does the drop down to filter the files shown in the Browse dialog need updating (on the fitting tab)? I was wondering if it needs a "texture" entry now in addition to bank 1, bank 2. Maybe not essential

DannyHindson reviewed (commented):
I've just remembered one other thing about this change. When I was looking into the problems you were having with 2-3 of the unit tests failing I noticed that a tearDown method I'd added in the engineering_diffraction_io_test wasn't behaving. This is part of the DecoderTest:
```
    def tearDown(self):
        self.gui.close()
```

Sometimes if the decode action has failed there isn't a "gui" attribute created on self at all by the time the code gets here. This tearDown method then raises its own exception making it hard to see what actually went wrong. If you have a minute, please could you had an `if hasattr(self, 'gui')` around this?

RichardWaiteSTFC committed:
Add check for gui attribute in io test teardown

RichardWaiteSTFC committed:
Hard code path to xml in CalInfo and don't change user dir on UI open

RichardWaiteSTFC committed:
Add group descriptors to calibration subplot titles

RichardWaiteSTFC committed:
Add release notes

RichardWaiteSTFC committed:
Update interface and testing docs/source/images/Capture

RichardWaiteSTFC committed:
Merge branch 'main' into 31905_add_texture_grouping_ENGINX_EngDiffUI

DannyHindson reviewed (changes requested):
I've added a few minor comments\requests but think we're pretty close now.

It's got quite a lot of commits. We should perhaps ensure these get squashed before it gets merged. The gatekeeper can do that or you can do on a rebase

RichardWaiteSTFC committed:
Remove additional runs for no archive access

Post #32756 these files will also be in the systems test data

RichardWaiteSTFC committed:
Avoid creating temp CalInfo obj to make prm name for indiv banks

Necessary when both banks are calibrated - it writes a prm for each bank individually

RichardWaiteSTFC committed:
Add method to CalInfo to make calibration subplot title

DannyHindson reviewed (approved):
Looks good thanks.

Please can a gatekeeper squash the commits on this PR when merging?
